### PR TITLE
improved extrinsic payment options interface (#244)

### PR DESCRIPTION
**Pull request by eupston**
@holygits here's the initial improvement to the interface as discuss we may want to pass the instance of the API object down to the extrinsic in another PR which would give us quite a lot of flexibility and potential for more abstraction.

### CODE REVIEW

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        return '';
+      }, toU8a(isBare?: BareOpts): Uint8Array {
+        return undefined;
+      },
+        extrinsic : this,
+        userFeeAssetId: feeAssetId,
+        maxPayment: _maxPayment
+    }
+    const getEstimatedFee = async () => {
+      const estimatedFee = await api.derive.fees.estimateFee(estimateFeeParams)
```
Comment (holygits) [inline packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:90]: so if I understand correctly `EstimateFeeParams` doesn't impl Codec so we implemented it here? I think we can relax that requirement in the types and already did so on the `master` branch cc @KarishmaBothara 
should make this more readable

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+      return res as SignatureOptions;
+    };
+    const mergedSignatureOpts = mergeDefinedObjects(options, this.signaturePayloadOptions);
+    return super.sign(account, mergedSignatureOpts);
+  }
+
+  /**
+   * @description Adds payment options such as slippage and tip to the Extrinsic signature
+   */
+  paymentOpts(api: Api, feeAssetId: number | string, slippage = 0, tip = 0): this{
```
Comment (holygits) [inline packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:67]: let's call it `setPaymentOpts` for consistency
also providing de-structuring will help for readability

```suggestion
  setPaymentOpts(api: Api, { feeAssetId: number | string, slippage = 0, tip = 0 }): this{
```

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      const estimatedFee = await api.derive.fees.estimateFee(estimateFeeParams)
+      return estimatedFee;
+    }
+    getEstimatedFee()
+      .then((estimatedfee) => {
+      const maxPayment = parseFloat(estimatedfee.toString()) * (1 + slippage)
+      const assetId = this.registry.createType('AssetId', feeAssetId);
+      const feeExchange = this.registry.createType('FeeExchange', { assetId, maxPayment }, 0);
+      const transactionPayment = this.registry.createType('ChargeTransactionPayment', {tip: tip, feeExchange});
+      if ('transactionPayment' in this.signaturePayloadOptions) {
```
Comment (holygits) [inline packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:99]: wondering when is this not `true` (genuine question)?

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        userFeeAssetId: feeAssetId,
+        maxPayment: _maxPayment
+    }
+    const getEstimatedFee = async () => {
+      const estimatedFee = await api.derive.fees.estimateFee(estimateFeeParams)
+      return estimatedFee;
+    }
+    getEstimatedFee()
+      .then((estimatedfee) => {
+      const maxPayment = parseFloat(estimatedfee.toString()) * (1 + slippage)
```
Comment (holygits) [inline packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:95]:
```suggestion
      const maxPayment = parseFloat(estimatedfee.toString()) * (1 + slippage);
```

**Review by holygits — COMMENTED** [review]
nice. general direction lgtm. thinking we just need a bunch of tests to check the compatibility with the current `signAndSend` implementation e.g.

`tx.setNonce(nonce).signAndSend(nonce)` <- what should be our default behaviour

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts (hunk 2 above)
Comment (eupston) [reply packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:67]: yup sure can change

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts (hunk 1 above)
Comment (eupston) [reply packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:90]: yeah definitely so `EstimateFeeParams` shouldn't need to extend `Codec` then? looks like it was added for Prerelease/2.0.0 alpha.2 (#299)

Comment (holygits) [reply packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:90]: yeah for `derive` queries arguments don't need to have `Codec` so we decided to relax that typing

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts (hunk 3 above)
Comment (eupston) [reply packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:99]: oo just when I'm using one object to store both `SignatureOptions | ExtrinsicPayloadValue` and `ExtrinsicPayloadValue` doesn't contain transactionPayment, but thinking about it now might be cleaner just to split out the signature and payload into two objects.

File: packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts (hunk 1 above)
Comment (eupston) [reply packages/types/src/interfaces/extrinsic/v1/Extrinsic.ts:90]: okay sweet i'll change that

**Comment by eupston** [comment]
> nice. general direction lgtm. thinking we just need a bunch of tests to check the compatibility with the current `signAndSend` implementation e.g.
> 
> `tx.setNonce(nonce).signAndSend(nonce)` <- what should be our default behaviour

yup for sure I added two test variations so far, one with the nonce being passed into `signAndSign` directly and one setting it via chaining. I'd say chaining should be default? means less overhead for user. In scenario above, if user sets nonce via chain and then passes in the nonce again to signAndSend it will use the one passed in. You reckon we should have a test to ensure it does use the latter?

**Comment by eupston** [comment]
closing PR and reopening changes on master as integration tests on develop aren't compatible with the latest cennznet version 1.5.0.

**Pull request closed without merging**